### PR TITLE
Canonicalize expressions involving typed literals

### PR DESCRIPTION
**Pull request by martint**
The current IR doesn't have a way to represent typed constants, so
it encodes them as Cast(Literal). This change extends the canonicalization
of comparison and arithmetic expressions to take that into account.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java [hunk 1]
```diff
@@ -156,4 +157,14 @@ public Expression rewriteExtract(Extract node, Void context, ExpressionTreeRewri
             throw new UnsupportedOperationException("not yet implemented: " + node.getField());
         }
     }
+
+    private static boolean isLiteral(Expression expression)
+    {
+        // Current IR has no way to represent typed constants, so it encodes them as Cast(Literal)
+        if (expression instanceof Cast && ((Cast) expression).getExpression() instanceof Literal) {
+            return true;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: This is not sufficient for:

https://github.com/prestosql/presto/blob/1116011e9f9e0a5b54740b349c3476412d6afdba/presto-main/src/main/java/io/prestosql/sql/planner/LiteralEncoder.java#L134-L138
(DOUBLE, REAL ±infinity)

https://github.com/prestosql/presto/blob/1116011e9f9e0a5b54740b349c3476412d6afdba/presto-main/src/main/java/io/prestosql/sql/planner/LiteralEncoder.java#L204-L212
(anything Block-based, varbinary, json or anything Slice-based other than varchar, char and decimal)

File: presto-main/src/test/java/io/prestosql/sql/TestExpressionInterpreter.java [hunk 2]
```diff
@@ -1132,16 +1132,16 @@ public void testSimpleCase()
     @Test
     public void testCoalesce()
     {
-        assertOptimizedEquals("coalesce(2 * 3 * unbound_long, 1 - 1, null)", "coalesce(6 * unbound_long, 0)");
-        assertOptimizedEquals("coalesce(2 * 3 * unbound_long, 1.0E0/2.0E0, null)", "coalesce(6 * unbound_long, 0.5E0)");
+        assertOptimizedEquals("coalesce(unbound_long * (2 * 3), 1 - 1, null)", "coalesce(unbound_long * 6, 0)");
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/TestExpressionInterpreter.java]: why did you change the input too (`coalesce(2 * 3 * unbound_long, 1 - 1, null)`) ?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java [hunk 3]
```diff
@@ -53,7 +54,7 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
         {
             // if we have a comparison of the form <constant> <op> <expr>, normalize it to
             // <expr> <op-flipped> <constant>
-            if (node.getLeft() instanceof Literal && !(node.getRight() instanceof Literal)) {
+            if (isLiteral(node.getLeft()) && !(isLiteral(node.getRight()))) {
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: unnecessary parens
```suggestion
            if (isLiteral(node.getLeft()) && !isLiteral(node.getRight())) {
```

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java [hunk 4]
```diff
@@ -66,7 +67,7 @@ public Expression rewriteArithmeticBinary(ArithmeticBinaryExpression node, Void
             if (node.getOperator() == MULTIPLY || node.getOperator() == ADD) {
                 // if we have a operation of the form <constant> [+|*] <expr>, normalize it to
                 // <expr> [+|*] <constant>
-                if (node.getLeft() instanceof Literal && !(node.getRight() instanceof Literal)) {
+                if (isLiteral(node.getLeft()) && !(isLiteral(node.getRight()))) {
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: unnecessary parens
```suggestion
                if (isLiteral(node.getLeft()) && !isLiteral(node.getRight())) {
```

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java [hunk 5]
```diff
@@ -156,4 +157,14 @@ public Expression rewriteExtract(Extract node, Void context, ExpressionTreeRewri
             throw new UnsupportedOperationException("not yet implemented: " + node.getField());
         }
     }
+
+    private static boolean isLiteral(Expression expression)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: `isConstant`?

**Review by findepi — APPROVED** [review]
LGTM, but can we do better, ie cover the encoded constants which escape Cast(Literal) pattern?

File: presto-main/src/test/java/io/prestosql/sql/TestExpressionInterpreter.java (hunk 2 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/sql/TestExpressionInterpreter.java]: I think constant folding is happening after canonicalization in some cases, so they weren't matching. I haven't investigated further, though.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java (hunk 3 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: heh... stupid IntelliJ refactoring.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java (hunk 5 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: Sure. Although isConstant implies more than what this is. It's not about arbitrary constant expressions, but AST-encoded typed literals.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java (hunk 1 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: I'm aware. I wanted to do the simplest possible thing that 1) doesn't require ExpressionInterpreter.optimize, 2) doesn't try to duplicate all the logic in LiteralEncoder, 3) covers a sufficient portion of the use cases that occur in practice

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java]: Sure. Please expand your comment. `// Current IR has no way to represent typed constants, so it encodes them as Cast(Literal` sounds like these other cases are being overlooked, rather than consciously ignored.

File: presto-main/src/test/java/io/prestosql/sql/TestExpressionInterpreter.java (hunk 2 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/sql/TestExpressionInterpreter.java]: Oh, yeah, `assertOptimizedEquals` is optimizing the expression directly.
